### PR TITLE
Move renorm to C++ and expose cumsum

### PR DESCRIPTION
**Pull request by colesbury**
Also allow cumprod forward in C++.

### CODE REVIEW

File: tools/autograd/templates/Functions.cpp [hunk 1]
```diff
@@ -165,6 +165,31 @@ Tensor mm_mat2_backward(const Tensor & grad, const Tensor & mat1, IntList sizes,
...
+  auto self_flat = flatten(self);
+  auto grad_flat = flatten(grad);
+  auto norm_flat = self_flat.norm(p, 1, true);
+  auto grad_output = (self_flat * grad_flat).sum(1, true);
+  auto nb = norm_backward(grad_output, self_flat, p, norm_flat, 1, true);
+  auto invnorm = (norm_flat + 1e-7).reciprocal();
+  auto grad_norm = unflatten(maxnorm * invnorm * (grad_flat - invnorm * nb));
+  auto norm = unflatten(norm_flat.expand_as(self_flat));
+
+  auto mask = Variable(norm < maxnorm).detach();
```
Comment (killeent) [inline tools/autograd/templates/Functions.cpp]: Why the `detach` here?

**Review by killeent — APPROVED** [review]

File: tools/autograd/templates/Functions.cpp (hunk 1 above)
Comment (colesbury) [reply tools/autograd/templates/Functions.cpp]: This is a little weird. Without this, we end up trying to negate a `ByteTensor` in the double-backwards. `ByteTensor` is unsigned so the negation fails.

The problem is because the comparison `mask = norm < maxnorm` currently has the result have `requires_grad=True`. Later on we do `1 - mask`, which is where the negation in the double-backwards comes in.

Comment (colesbury) [reply tools/autograd/templates/Functions.cpp]: I think adding something like TensorFlows `tf.where` would help here. We want a way to express element-wise `cond ? a : b` in Tensor operations. The current method of multiplications and `1 - cond` isn't ideal.

Comment (apaszke) [reply tools/autograd/templates/Functions.cpp]: Why does comparison has `requires_grad=True`?

Comment (apaszke) [reply tools/autograd/templates/Functions.cpp]: But yes, `np.where`-like op would be useful.

File: tools/autograd/derivatives.yaml [hunk 2]
```diff
@@ -161,7 +161,8 @@
   self: other.cross(grad, dim)
   other: grad.cross(self, dim)
 
-- name: cumprod  # complicated
+- name: cumprod(Tensor self, int64_t dim)
+  self: not_implemented("cumprod")  # implemented in Python
```
Comment (apaszke) [inline tools/autograd/derivatives.yaml]: This is confusing. Why do you add it here if it's going through Python anyway?

Comment (colesbury) [reply tools/autograd/derivatives.yaml]: Two reasons:

1) It's useful for people using the C++ interface
2) Once we merge Variable and Tensor, Python-implemented autograd functions will operate on Variables in the forward() (but with gradients disabled). The C++ cumprod needs to be exposed so that the Python-implemented autograd function can call it.

File: tools/autograd/templates/Functions.cpp (hunk 1 above)
Comment (colesbury) [reply tools/autograd/templates/Functions.cpp]: I have an issue tracking it here:
https://github.com/pytorch/pytorch/issues/3348

We should fix it, but I think we'll also need to change some tests. IIRC, a bunch of them will fail because the backward engine throws an exception if things don't require grad.

File: tools/autograd/derivatives.yaml (hunk 2 above)
Comment (apaszke) [reply tools/autograd/derivatives.yaml]: I understand the first point, but I can't see how the second would work. We would need to use the C++ function in no grad mode in Python, and Python implementation in all other cases.

Comment (colesbury) [reply tools/autograd/derivatives.yaml]: I was thinking it might look something like:

```python
class Cumprod(Function):
    @staticmethod
    def forward(ctx, input, dim):
        ctx.dim = dim
        ctx.save_for_backward(input)
        return torch._C._VariableBase.cumprod(input, ctx.dim)
  
    @staticmethod
    def backward(ctx, grad_output):
        ...
```
